man/man1/cat2bsq.1
.TH cat2bsq 1  "USER COMMANDS"
.SH NAME
cat2bsq \- Concatenates output files from distributed LPJmL simulations.
.SH SYNOPSIS
.B cat2bsq
[-type {float|double|short|int|char}] [-o \fIoutfile\fP] [-bands n] \fIinfilename\fP ...
.SH DESCRIPTION
Program concatenates output files from the LPJmL model to one output file. It used to join output files from
.B
lpjdistribute
and
.B lpjdistribute_river
scripts.  Program is obsolete for the MPI version of \fBlpjml\fP.
.SH OPTIONS
.TP
\-type {float|double|short|int|char}
Datatype of output. Default is float.
.TP
\-bands n
Number of bands. Default is 103.
.TP
\-o \fIoutfile\fP
Set filename of the concatenated output file. Default is output to stdout.
.TP
.I infilename
Name of output file(s) to join.
.SH EXAMPLE
.TP
Concatenate output files to the file \fIvegc.bin\fP:
.B cat2bsq
\-o vegc.bin vegc_*.bin
.PP

.SH EXIT STATUS
.B cat2bsq
returns a zero exit status if the output files have been concatenated.
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
lpjml(1), lpjcat(1), output_bsq(1)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)

### Changed

- Axis names and missing values for NetCDF output are read from JSON file `netcdf.cjson`.
- Warnings in `addheader` replaced by terminating errors.
- Function `openinputfile()` returns an error if cell size differs in JSON metafile.
- Function `openmetafile()` returns an error if binary file is empty.

### Added

- Environment variables `LPJNOPP` and `LPJPEDANTIC` added. If set to `true`, preprocessor is disabled and pedantic checking is enabled, respectively.
- Info about `--help` option added in case of error in command line arguments.
- Option `-config` added to `bin2cdf`, `country2cdf` and `clm2cdf` in order to read axis names and missing values from a JSON file. File `netcdf.json` added as a template.
- Option `-json` added to `arr2clm`, `asc2clm`, `txt2grid`, `txt2clm`, `getcountry`, and `cdf2coord` in order to create additional JSON metafiles.
- Option `-inpath` added to `regridlpj` in order to set directory where data can be found.
- Option `-double` added to `txt2grid` in order to create grid files with coordinate values of type double.
- Option `-timestep` added to `addheader`.
- Check for correct number of time steps added in `openclimate_netcdf.c`.
- Option `sum`, `tsum` and `tmean` to `mathclm` added.


## [5.9.15] - 2024-09-30

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Sibyll Schaphoff (sibylls@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de)

### Added

- Option `-scale` added to `bin2cdf` in order to scale output.
- Option `-notime` added to `bin2cdf` in order to omit time axis.
- `time_bnds`, `lat_bnds`, and `lon_bnds` arrays added to NetCDF output for lpjml runs and the `bin2cdf` utility.
- Check for identical output filenames added in `fscanoutput.c`. 

### Changed

- If `"with_days"` is set to true in the lpjml configuration file, time axis of yearly output is also set in units of days instead of years.
- Time in NetCDF output is always set in the middle between the lower and upper time boundaries.

### Removed

- Unused file `cflux_sum.c` removed.
- Conversion of obsolete file `manage.par` removed from utility `manage2js`.

### Fixed

- Separate output for each year by setting `"name" : "filename_%d"` is now written correctly.
- Number of bands set to 1 in metafile of grid NetCDF output.
- Name of longitude/latitude variables correctly derived in `getlatlon_netcdf.c`.
- Misspelled option `-netdcdf4` corrected to `-netcdf4` in `clm2cdf.c`.
- `lpj_climber4.c` updated to compile without errors.
- Writing different variables into one NetCDF file fixed in `create_netcdf.c`. Only NetCDF outputs with one band can be written to one file.


## [5.9.14] - 2024-09-30

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de)
- code review: Sibyll Schaphoff (sibylls@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)

### Fixed

- `pft->vscal` not set to `NDAYEAR` in `update_annual.c`. (issue #364).
- `freezefrac2soil.c` corrected to avoid division by zero.
- size of option array for `sowing_date_option` and `crop_phu_option` now correctly specified in `fscanconfig.c`.


## [5.9.13] - 2024-09-26

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Sebastian Ostberg (ostberg@pik-potsdam.de), Marie Hemmen (hemmen@pik-potsdam.de)

### Changed

- `USE_NETCDF4` compile option replaced by `"netcdf4"` boolean flag in the LPJmL configuration file. If set compression of NetCDF4 files can be enabled and PFT names are written as strings instead of character arrays.
- Formatting of man pages harmonized.
- If lpjml is compiled without `-DUSE_NETCDF` flag then `fscanconfig.c` and `fscanoutput.c` return with an error for NetCDF files specified for input/output.

### Added

- Option `-netcdf4` added to `bin2cdf`, `clm2cdf`, `country2cdf`, and `drainage2cdf` utility to enable NetCDF4 format.
- Boolean flag `"netcdf4"` added to LPJmL configuration file in order to enable NetCDF4 format.
- Option `-compress` added to `drainage2cdf` utility to enable file compression.
- Man page for `drainage2cdf` added.
- Check for valid compression value added for lpjml and utilities.
- Missing option `"global_netcdf"` added to LPJmL config file.

### Removed

- Obsolete man page for `writeregioncode()` removed.

### Fixed

- Missing deallocation of memory added in `cpl_init.c`, `fscanlandcovermap.c`, `fscanoutput.c`, `writearea.c`, `create_pft_netcdf.c`, `newgrid.c`, `fscanagtreemap.c`, `celldata.c` in case of error.
- Syntax error fixed in `update_daily.c` in IMAGE coupling.
- Typo in `README` corrected.
- `establishmentpft.c`, `cdf2coord.c` and `cdf2grid.c` modified to compile without warnings using gcc.
- Debug flag in `send_token_coupler.c` and `openoutput_coupler.c` corrected to `DEBUG_COUPLER`.
- Code changed to compile without errors/warnings for `-DIMAGE -DCOUPLED` setting.
- Calculation of offsets and counts corrected in `readintdata_netdf.c` and missing loop over bands added.


## [5.9.12] - 2024-09-24

### Contributors

- author: Christoph Müller (cmueller@pik-potsdam.de)
- code review: Werner von Bloh (bloh@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de)

### Added


src/utils/cat2bsq.c
/**************************************************************************************/
/**                                                                                \n**/
/**                      c a t 2 b s q  .  c                                       \n**/
/**                                                                                \n**/
/**     Program  concatenates  output files from the LPJmL model                   \n**/
/**     to one output file.                                                        \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <string.h>
#include <errno.h>
/*#include <sys/types.h>*/
#include <sys/stat.h>
#include "types.h"
#include "errmsg.h"

#define USAGE "Usage: cat2bsq [-type {byte|short|int|float|double}] [-o outfile] [-bands n] infilename ...\n"

int main(int argc,char **argv)
{
  FILE *ofp;
  FILE **ifp;
  void *data;
  int i,j,k,p,*n,runs,bands,sum;
  char *outfile,*endptr;
  struct stat filestat;
  k=LPJ_FLOAT;
  bands=103;
  outfile=NULL;
  for(i=1;i<argc;i++)
    if(argv[i][0]=='-')
    {
      if(!strcmp(argv[i],"-type"))
      {
        if(argc==i+1)
        {
          fprintf(stderr,"Argument missing.\n");
          fprintf(stderr,USAGE);
          return EXIT_FAILURE;
        }
        else
        {
          k=findstr(argv[++i],typenames,5);
          if(k==NOT_FOUND)
          {
            fprintf(stderr,"Invalid argument '%s' for option '-type'.\n",argv[i]);
            fprintf(stderr,USAGE);
            return EXIT_FAILURE;
          }
        }
      }
      else if(!strcmp(argv[i],"-o"))
      {
        if(argc==i+1)
        {
          fprintf(stderr,"Argument missing.\n");
          fprintf(stderr,USAGE);
          return EXIT_FAILURE;
        }
        else
          outfile=argv[++i];
      }
      else if(!strcmp(argv[i],"-bands"))
      {
        if(argc==i+1)
        {
          fprintf(stderr,"Argument missing.\n");
          fprintf(stderr,USAGE);
          return EXIT_FAILURE;
        }
        else
        {
          bands=strtol(argv[++i],&endptr,10);
          if(*endptr!='\0')
          {
            fprintf(stderr,"Invalid number '%s' for option '-bands'.\n",argv[i]);
            return EXIT_FAILURE;
          }
        }
        if(bands<1)
        {
          fprintf(stderr,"Number of bands=%d must be greater than zero.\n",bands);
          return EXIT_FAILURE;
        }
      }
      else
      {
        fprintf(stderr,"Invalid option '%s'.\n",argv[i]);
        fprintf(stderr,USAGE);
        return EXIT_FAILURE;
      }
    }
    else
      break;
  runs=argc-i;
  if(runs==0)
  {
    fprintf(stderr,"Filename is missing.\n" USAGE);
    return EXIT_FAILURE;
  }
  if(outfile==NULL)
    ofp=stdout;
  else
  {
    ofp=fopen(outfile,"wb");
    if(ofp==NULL)
    {
      fprintf(stderr,"Error opening '%s': %s\n",outfile,strerror(errno));
      return EXIT_FAILURE;
    }
  }
  ifp=newvec(FILE *,runs);
  n=(int *)malloc(runs*sizeof(int));
  check(n);
  for(j=0;j<runs;j++)
  {

    if(!(ifp[j]=fopen(argv[i+j],"rb")))
    {
      fprintf(stderr,"Error: File open failed on path %s: %s\n",argv[i+j],strerror(errno));
      return EXIT_FAILURE;
    }
    fstat(fileno(ifp[j]),&filestat);
    if(filestat.st_size % (bands*typesizes[k]))
    {
      fprintf(stderr,"Error: Filesize=%zd is not divisible without remainder=%d, bands=%d, size=%d.",
           filestat.st_size,(int)(filestat.st_size % (bands*typesizes[k])),bands,(int)typesizes[k]);
      return EXIT_FAILURE;
    }
    /* calculate number of pixels in each chunk and store it in array n[i] */
    n[j]=filestat.st_size/bands/typesizes[k];
  }

  for(p=0;p<bands;p++)
  {
    for(j=0;j<runs;j++)
    {
      if(outfile!=NULL && p==0) 
        printf("run: %d n: %d\n",i,n[j]);
      fseek(ifp[j],typesizes[k]*n[j]*p,SEEK_SET);
      data=malloc(n[j]*typesizes[k]);
      if(data==NULL)
      {
        printallocerr("data");
        return EXIT_FAILURE;
      }
      if(fread(data,typesizes[k],n[j],ifp[j])!=1){
        fprintf(stderr,"Error reading %s.\n",argv[i+j]);
        exit(1);
      }
      fwrite(data,typesizes[k],n[j],ofp);
      free(data);
    }
  }
  if(outfile!=NULL)
  {
    sum=0;
    for(j=0;j<runs;j++)
      sum+=n[j];
    printf("Number of cells: %d\n",sum); 
  }
  for(j=0;j<runs;j++)
    fclose(ifp[j]);
  fclose(ofp);
  return EXIT_SUCCESS;
} /* of 'main' */
